src/sentry/eventstream/kafka.py

```diff
@@ -20,31 +19,37 @@
 EVENT_PROTOCOL_VERSION = 1
 
 
-class KafkaPublisher(object):
-    def __init__(self, connection):
-        self.connection = connection or {}
+class KafkaEventStream(EventStream):
+    def __init__(self, publish_topic='events', producer_configuration=None, **options):
+        if producer_configuration is None:
+            producer_configuration = {}
 
-    @cached_property
-    def client(self):
-        return KafkaProducer(**self.connection)
+        self.publish_topic = publish_topic
+        self.producer_configuration = producer_configuration
 
-    def publish(self, topic, value, key=None):
-        return self.client.send(topic, key=key, value=value)
+    @cached_property
+    def producer(self):
+        return Producer(**self.producer_configuration)
 
-
-class KafkaEventStream(EventStream):
-    def __init__(self, publish_topic='events', sync=False, connection=None, **options):
-        self.publish_topic = publish_topic
-        self.pubsub = KafkaPublisher(connection)
-        if not sync:
-            self.pubsub = QueuedPublisher(self.pubsub)
+    def delivery_callback(self, error, message):
+        logger.warning('Could not publish event (error: %s): %r', error, message)
 
     def publish(self, group, event, is_new, is_sample, is_regression, is_new_group_environment, primary_hash, skip_consume=False):
         project = event.project
         retention_days = quotas.get_event_retention(
             organization=Organization(project.organization_id)
         )
 
+        # Polling the producer is required to ensure callbacks are fired. This
+        # means that the latency between a message being delivered (or failing
+        # to be delivered) and the corresponding callback being fired is
+        # roughly the same as the duration of time that passes between publish
+        # calls. If this ends up being too high, the publisher should be moved
+        # into a background thread that can poll more frequently without
+        # interfering with request handling. (This does `poll` does not act as
+        # a heartbeat for the purposes of any sort of session expiration.)
+        self.producer.poll(0.0)
```

Comment:
The process running this code is doing a tight-loop that's processing events, right? If so, I think the `poll` is probably fine. But if this ever moved to happen in a web process that could `publish` and then do 30 minutes of random API requests it would get pretty weird and I'd agree with the extra thread.

Reply:
I think the webs are handling both ingest as well as API, unless anything has changed recently?

Reply:
I thought webs did that first-pass method, then we did the whole Rabbit/Redis dance, and actual event processing (so including `eventstream` publish) happened on Celery workers?

Reply:
Oops, yeah — you're right, ignore me